

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>TOPMed Demo &mdash; BioNeuralNet 1.0.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/nbsphinx-code-cells.css?v=2aa19091" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=71272d9f"></script>
      <script src="_static/doctools.js?v=9bcbadda"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
      <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
      <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Tutorials" href="tutorials/index.html" />
    <link rel="prev" title="TCGA-BRCA Dataset" href="TCGA-BRCA_Dataset.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            BioNeuralNet
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="gnns.html">GNN Embeddings</a></li>
<li class="toctree-l1"><a class="reference internal" href="clustering.html">Correlated Clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="metrics.html">Metrics</a></li>
<li class="toctree-l1"><a class="reference internal" href="utils.html">Utils</a></li>
<li class="toctree-l1"><a class="reference internal" href="downstream_tasks.html">Downstream Tasks</a></li>
<li class="toctree-l1"><a class="reference internal" href="Quick_Start.html">BioNeuralNet Demo: Quick Start Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="TCGA-BRCA_Dataset.html">TCGA-BRCA Dataset</a></li>
<li class="toctree-l1"><a class="reference internal" href="TCGA-BRCA_Dataset.html#Preprocessing">Preprocessing</a></li>
<li class="toctree-l1"><a class="reference internal" href="TCGA-BRCA_Dataset.html#Optional:-Load-the-data-we-just-saved-to-make-sure-it-looks-okay.">Optional: Load the data we just saved to make sure it looks okay.</a></li>
<li class="toctree-l1"><a class="reference internal" href="TCGA-BRCA_Dataset.html#Easy-Access-via-DatasetLoader">Easy Access via DatasetLoader</a></li>
<li class="toctree-l1"><a class="reference internal" href="TCGA-BRCA_Dataset.html#Preparing-Multi-Omics-Data-for-downstream-tasks">Preparing Multi-Omics Data for downstream tasks</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">TOPMed Demo</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#Trans-Omics-for-Precision-Medicine-|-TOPMed">Trans-Omics for Precision Medicine | TOPMed</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Live-Presentation-of-BioNeuralNet.">Live Presentation of BioNeuralNet.</a></li>
<li class="toctree-l3"><a class="reference internal" href="#BioNeuralNet-Components">BioNeuralNet Components</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#Loading-your-data:">Loading your data:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Ease-of-component-exploration-via-DatasetLoader">Ease of component exploration via <code class="docutils literal notranslate"><span class="pre">DatasetLoader</span></code></a></li>
<li class="toctree-l2"><a class="reference internal" href="#Generating-a-Multi-Omics-Network-using-SmCCNet">Generating a Multi-Omics Network using SmCCNet</a></li>
<li class="toctree-l2"><a class="reference internal" href="#SmCCNet-Output">SmCCNet Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Generating-Low-Dimensional-Embeddings-using-Graph-Neural-Networks-to-capture-meaningful-biological-interactions.">Generating Low-Dimensional Embeddings using Graph Neural Networks to capture meaningful biological interactions.</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Output">Output</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Embeddings-visualization">Embeddings visualization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Using-the-Embeddings">Using the Embeddings</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Comparing-results">Comparing results</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Network-Visulization">Network Visulization</a></li>
<li class="toctree-l2"><a class="reference internal" href="#Correlated-Clustering">Correlated Clustering</a></li>
<li class="toctree-l2"><a class="reference internal" href="#DPMON-(Disease-Prediction-using-Multi-Omics-Networks)-reuses-the-same-GNN-architectures-but-with-a-different-objective:">DPMON (Disease Prediction using Multi-Omics Networks) reuses the same GNN architectures but with a different objective:</a></li>
<li class="toctree-l2"><a class="reference internal" href="#DPMON-allows-BioNeuralNet-users-to-significatly-improve-phenotype-predictions-with-a-few-lines-of-code.">DPMON allows BioNeuralNet users to significatly improve phenotype predictions with a few lines of code.</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="tutorials/index.html">Tutorials</a></li>
<li class="toctree-l1"><a class="reference internal" href="external_tools/index.html">External Tools</a></li>
<li class="toctree-l1"><a class="reference internal" href="user_api.html">User API</a></li>
<li class="toctree-l1"><a class="reference internal" href="faq.html">Acknowledgments</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">BioNeuralNet</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">TOPMed Demo</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/TOPMED.ipynb.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="TOPMed-Demo">
<h1>TOPMed Demo<a class="headerlink" href="#TOPMed-Demo" title="Link to this heading"></a></h1>
<section id="Trans-Omics-for-Precision-Medicine-|-TOPMed">
<h2>Trans-Omics for Precision Medicine | TOPMed<a class="headerlink" href="#Trans-Omics-for-Precision-Medicine-|-TOPMed" title="Link to this heading"></a></h2>
<section id="Live-Presentation-of-BioNeuralNet.">
<h3>Live Presentation of BioNeuralNet.<a class="headerlink" href="#Live-Presentation-of-BioNeuralNet." title="Link to this heading"></a></h3>
<p>A <strong>step-by-step</strong> guide to <strong>BioNeuralNet</strong>.</p>
<ul class="simple">
<li><p>This demonstration was made specifically for the 2025 TOPMed Annual Meeting. Featuring artificial intelligence and machine learning.</p></li>
<li><p>For more information on TOPMed and their mission, please visit <a class="reference external" href="https://topmed.nhlbi.nih.gov/">TOPMed</a>.</p></li>
<li><p>This presentation uses an outdated version of <code class="docutils literal notranslate"><span class="pre">BioNeuralNet</span></code>, please refer to <code class="docutils literal notranslate"><span class="pre">TCGA-BRCA_Dataset.ipynb</span></code> for an up to date end-to-end example.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># In terminal use:</span>
<span class="n">pip</span> <span class="n">install</span> <span class="n">bioneuralnet</span>

<span class="c1"># In Jupiter notebook use</span>
<span class="o">!{</span>sys.executable<span class="o">}</span><span class="w"> </span>-m<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>bioneuralnet
</pre></div>
</div>
</div>
</section>
<section id="BioNeuralNet-Components">
<h3>BioNeuralNet Components<a class="headerlink" href="#BioNeuralNet-Components" title="Link to this heading"></a></h3>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.network_embedding</span><span class="w"> </span><span class="kn">import</span> <span class="n">GNNEmbedding</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.downstream_task</span><span class="w"> </span><span class="kn">import</span> <span class="n">SubjectRepresentation</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.downstream_task</span><span class="w"> </span><span class="kn">import</span> <span class="n">DPMON</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.clustering</span><span class="w"> </span><span class="kn">import</span> <span class="n">CorrelatedPageRank</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.clustering</span><span class="w"> </span><span class="kn">import</span> <span class="n">CorrelatedLouvain</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.clustering</span><span class="w"> </span><span class="kn">import</span> <span class="n">HybridLouvain</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.datasets</span><span class="w"> </span><span class="kn">import</span> <span class="n">DatasetLoader</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">omics_correlation</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">cluster_correlation</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">louvain_to_adjacency</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">evaluate_rf</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_variance_distribution</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_variance_by_feature</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_performance</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_embeddings</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_network</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">compare_clusters</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">rdata_to_df</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">get_logger</span>

<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.external_tools</span><span class="w"> </span><span class="kn">import</span> <span class="n">SmCCNet</span>
</pre></div>
</div>
</div>
</section>
</section>
<section id="Loading-your-data:">
<h2>Loading your data:<a class="headerlink" href="#Loading-your-data:" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>If you data is stored in a csv file, it can be loaded by following the example below.</p></li>
<li><p>After loading your data, the remaining steps will be the same.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>

<span class="n">genomic</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span> <span class="s2">&quot;genes.csv&quot;</span><span class="p">)</span>
<span class="n">proteomics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;proteins.csv&quot;</span><span class="p">)</span>
<span class="n">metabolites</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;metabolites.csv&quot;</span><span class="p">)</span>
<span class="n">phenotype</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;phenotype.csv&quot;</span><span class="p">)</span>
<span class="n">clinical_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s2">&quot;clinical.csv&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Ease-of-component-exploration-via-DatasetLoader">
<h2>Ease of component exploration via <code class="docutils literal notranslate"><span class="pre">DatasetLoader</span></code><a class="headerlink" href="#Ease-of-component-exploration-via-DatasetLoader" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>This component allows users to explore BioNeuralNet capabilities.</p></li>
<li><p>DatasetLoader <code class="docutils literal notranslate"><span class="pre">example1</span></code> is synthetic and purely for example purposes.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.datasets</span><span class="w"> </span><span class="kn">import</span> <span class="n">DatasetLoader</span>
<span class="n">loader</span> <span class="o">=</span> <span class="n">DatasetLoader</span><span class="p">(</span><span class="s2">&quot;example1&quot;</span><span class="p">)</span>
<span class="n">omics1</span><span class="p">,</span> <span class="n">omics2</span><span class="p">,</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">clinical</span> <span class="o">=</span> <span class="n">loader</span><span class="o">.</span><span class="n">load_data</span><span class="p">()</span>

<span class="n">display</span><span class="p">(</span><span class="n">omics1</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">omics2</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">phenotype</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">clinical</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene_1</th>
      <th>Gene_2</th>
      <th>Gene_3</th>
      <th>Gene_4</th>
      <th>Gene_5</th>
      <th>Gene_6</th>
      <th>Gene_7</th>
      <th>Gene_8</th>
      <th>Gene_9</th>
      <th>Gene_10</th>
      <th>...</th>
      <th>Gene_491</th>
      <th>Gene_492</th>
      <th>Gene_493</th>
      <th>Gene_494</th>
      <th>Gene_495</th>
      <th>Gene_496</th>
      <th>Gene_497</th>
      <th>Gene_498</th>
      <th>Gene_499</th>
      <th>Gene_500</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Samp_1</th>
      <td>22.485701</td>
      <td>40.353720</td>
      <td>31.025745</td>
      <td>20.847206</td>
      <td>26.697293</td>
      <td>30.205449</td>
      <td>23.512005</td>
      <td>33.677622</td>
      <td>19.430333</td>
      <td>30.260153</td>
      <td>...</td>
      <td>12.967210</td>
      <td>13.334132</td>
      <td>13.262070</td>
      <td>13.608147</td>
      <td>13.338606</td>
      <td>15.247613</td>
      <td>13.603157</td>
      <td>14.727795</td>
      <td>13.400950</td>
      <td>12.769172</td>
    </tr>
    <tr>
      <th>Samp_2</th>
      <td>37.058850</td>
      <td>34.052233</td>
      <td>33.487020</td>
      <td>23.531461</td>
      <td>26.754628</td>
      <td>31.735945</td>
      <td>22.795952</td>
      <td>29.301536</td>
      <td>14.936397</td>
      <td>30.823015</td>
      <td>...</td>
      <td>11.886247</td>
      <td>13.241386</td>
      <td>12.445773</td>
      <td>13.400288</td>
      <td>12.587993</td>
      <td>15.048676</td>
      <td>12.018609</td>
      <td>14.513958</td>
      <td>12.066379</td>
      <td>12.583460</td>
    </tr>
    <tr>
      <th>Samp_3</th>
      <td>20.530767</td>
      <td>31.669623</td>
      <td>35.189567</td>
      <td>20.952544</td>
      <td>25.018826</td>
      <td>32.157235</td>
      <td>25.069464</td>
      <td>22.853719</td>
      <td>18.220225</td>
      <td>23.092805</td>
      <td>...</td>
      <td>12.198357</td>
      <td>12.869640</td>
      <td>12.841290</td>
      <td>13.127437</td>
      <td>12.607377</td>
      <td>14.398177</td>
      <td>12.554904</td>
      <td>14.339382</td>
      <td>12.891962</td>
      <td>12.760553</td>
    </tr>
    <tr>
      <th>Samp_4</th>
      <td>33.186888</td>
      <td>38.480880</td>
      <td>18.897097</td>
      <td>31.823300</td>
      <td>34.049383</td>
      <td>38.799887</td>
      <td>24.106468</td>
      <td>12.397175</td>
      <td>13.724255</td>
      <td>27.703085</td>
      <td>...</td>
      <td>12.197502</td>
      <td>13.312536</td>
      <td>11.645099</td>
      <td>13.933684</td>
      <td>12.103087</td>
      <td>15.133432</td>
      <td>12.436027</td>
      <td>15.116888</td>
      <td>12.810732</td>
      <td>12.972879</td>
    </tr>
    <tr>
      <th>Samp_5</th>
      <td>28.961981</td>
      <td>41.060494</td>
      <td>28.494956</td>
      <td>18.374495</td>
      <td>30.815238</td>
      <td>24.004535</td>
      <td>29.616296</td>
      <td>24.364045</td>
      <td>11.409338</td>
      <td>33.599828</td>
      <td>...</td>
      <td>14.157141</td>
      <td>13.386978</td>
      <td>13.007109</td>
      <td>13.826532</td>
      <td>12.343930</td>
      <td>15.280175</td>
      <td>13.363962</td>
      <td>14.008675</td>
      <td>12.479124</td>
      <td>12.156407</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Samp_354</th>
      <td>24.520652</td>
      <td>28.595409</td>
      <td>31.299666</td>
      <td>32.095379</td>
      <td>33.659730</td>
      <td>29.353948</td>
      <td>29.871659</td>
      <td>25.089732</td>
      <td>14.437588</td>
      <td>27.261916</td>
      <td>...</td>
      <td>11.748228</td>
      <td>13.096453</td>
      <td>12.768936</td>
      <td>13.228560</td>
      <td>11.595884</td>
      <td>15.251601</td>
      <td>12.055375</td>
      <td>15.530606</td>
      <td>13.644383</td>
      <td>13.018032</td>
    </tr>
    <tr>
      <th>Samp_355</th>
      <td>31.252789</td>
      <td>28.988087</td>
      <td>29.574195</td>
      <td>31.189288</td>
      <td>32.098841</td>
      <td>26.849138</td>
      <td>22.619669</td>
      <td>28.006670</td>
      <td>8.615549</td>
      <td>30.838843</td>
      <td>...</td>
      <td>11.460876</td>
      <td>12.994504</td>
      <td>13.325608</td>
      <td>13.386970</td>
      <td>12.335143</td>
      <td>14.966097</td>
      <td>13.998975</td>
      <td>13.250821</td>
      <td>12.947672</td>
      <td>13.161434</td>
    </tr>
    <tr>
      <th>Samp_356</th>
      <td>24.894826</td>
      <td>25.944887</td>
      <td>30.852641</td>
      <td>26.705158</td>
      <td>30.102546</td>
      <td>37.197952</td>
      <td>35.032309</td>
      <td>34.775576</td>
      <td>13.354611</td>
      <td>29.697348</td>
      <td>...</td>
      <td>12.884821</td>
      <td>12.785372</td>
      <td>12.954534</td>
      <td>13.906209</td>
      <td>12.074805</td>
      <td>15.519179</td>
      <td>12.742078</td>
      <td>14.445011</td>
      <td>12.129990</td>
      <td>13.844271</td>
    </tr>
    <tr>
      <th>Samp_357</th>
      <td>17.034337</td>
      <td>38.574705</td>
      <td>25.095201</td>
      <td>37.062442</td>
      <td>35.417758</td>
      <td>29.753145</td>
      <td>25.855997</td>
      <td>25.603043</td>
      <td>13.180794</td>
      <td>31.188598</td>
      <td>...</td>
      <td>13.278454</td>
      <td>13.445404</td>
      <td>12.457436</td>
      <td>14.036252</td>
      <td>13.412114</td>
      <td>14.158416</td>
      <td>13.284701</td>
      <td>13.662274</td>
      <td>12.943670</td>
      <td>13.996352</td>
    </tr>
    <tr>
      <th>Samp_358</th>
      <td>20.839167</td>
      <td>27.099788</td>
      <td>31.038453</td>
      <td>19.410859</td>
      <td>31.818995</td>
      <td>33.493625</td>
      <td>22.769369</td>
      <td>22.869481</td>
      <td>16.839248</td>
      <td>24.644476</td>
      <td>...</td>
      <td>12.519057</td>
      <td>13.308512</td>
      <td>12.261076</td>
      <td>11.211687</td>
      <td>12.058898</td>
      <td>15.349873</td>
      <td>13.166209</td>
      <td>15.295902</td>
      <td>13.257230</td>
      <td>13.178058</td>
    </tr>
  </tbody>
</table>
<p>358 rows × 500 columns</p>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Mir_1</th>
      <th>Mir_2</th>
      <th>Mir_3</th>
      <th>Mir_4</th>
      <th>Mir_5</th>
      <th>Mir_6</th>
      <th>Mir_7</th>
      <th>Mir_8</th>
      <th>Mir_9</th>
      <th>Mir_10</th>
      <th>...</th>
      <th>Mir_91</th>
      <th>Mir_92</th>
      <th>Mir_93</th>
      <th>Mir_94</th>
      <th>Mir_95</th>
      <th>Mir_96</th>
      <th>Mir_97</th>
      <th>Mir_98</th>
      <th>Mir_99</th>
      <th>Mir_100</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Samp_1</th>
      <td>15.223913</td>
      <td>17.545826</td>
      <td>15.784719</td>
      <td>14.891983</td>
      <td>10.348205</td>
      <td>9.689755</td>
      <td>11.093630</td>
      <td>13.189246</td>
      <td>12.526561</td>
      <td>10.311609</td>
      <td>...</td>
      <td>7.551462</td>
      <td>5.407878</td>
      <td>13.933045</td>
      <td>10.287521</td>
      <td>10.213316</td>
      <td>10.609973</td>
      <td>14.554996</td>
      <td>10.955650</td>
      <td>11.422531</td>
      <td>10.862970</td>
    </tr>
    <tr>
      <th>Samp_2</th>
      <td>16.306965</td>
      <td>16.672830</td>
      <td>13.361529</td>
      <td>14.488549</td>
      <td>12.660905</td>
      <td>11.333613</td>
      <td>11.254930</td>
      <td>13.457435</td>
      <td>13.279747</td>
      <td>10.428848</td>
      <td>...</td>
      <td>6.862413</td>
      <td>7.309226</td>
      <td>13.586180</td>
      <td>11.975544</td>
      <td>11.496937</td>
      <td>10.653742</td>
      <td>14.414225</td>
      <td>11.811004</td>
      <td>12.413667</td>
      <td>10.719110</td>
    </tr>
    <tr>
      <th>Samp_3</th>
      <td>16.545119</td>
      <td>16.735005</td>
      <td>14.617472</td>
      <td>17.845267</td>
      <td>13.822790</td>
      <td>11.329333</td>
      <td>11.171749</td>
      <td>12.715107</td>
      <td>13.787771</td>
      <td>10.364577</td>
      <td>...</td>
      <td>6.874958</td>
      <td>7.754733</td>
      <td>13.847029</td>
      <td>12.424403</td>
      <td>10.930177</td>
      <td>10.255484</td>
      <td>13.570352</td>
      <td>11.311925</td>
      <td>11.072915</td>
      <td>11.418794</td>
    </tr>
    <tr>
      <th>Samp_4</th>
      <td>13.986899</td>
      <td>16.207432</td>
      <td>16.293078</td>
      <td>17.725286</td>
      <td>12.300565</td>
      <td>9.844108</td>
      <td>11.586551</td>
      <td>11.878702</td>
      <td>14.594392</td>
      <td>10.936651</td>
      <td>...</td>
      <td>9.615623</td>
      <td>6.693593</td>
      <td>13.840728</td>
      <td>12.245466</td>
      <td>10.836894</td>
      <td>11.502232</td>
      <td>15.483399</td>
      <td>10.812811</td>
      <td>10.121957</td>
      <td>11.039089</td>
    </tr>
    <tr>
      <th>Samp_5</th>
      <td>16.338332</td>
      <td>17.393869</td>
      <td>16.397925</td>
      <td>15.853725</td>
      <td>13.387675</td>
      <td>10.599414</td>
      <td>12.355466</td>
      <td>12.450426</td>
      <td>13.778779</td>
      <td>10.405327</td>
      <td>...</td>
      <td>9.146865</td>
      <td>8.104206</td>
      <td>13.903452</td>
      <td>11.423350</td>
      <td>9.997107</td>
      <td>10.744586</td>
      <td>14.465583</td>
      <td>11.730166</td>
      <td>12.206151</td>
      <td>10.724849</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Samp_354</th>
      <td>15.065065</td>
      <td>16.079830</td>
      <td>14.635616</td>
      <td>17.013845</td>
      <td>11.612843</td>
      <td>9.850100</td>
      <td>11.375670</td>
      <td>12.870558</td>
      <td>13.873811</td>
      <td>11.369083</td>
      <td>...</td>
      <td>8.110691</td>
      <td>8.124900</td>
      <td>14.316491</td>
      <td>12.599296</td>
      <td>10.302683</td>
      <td>11.286541</td>
      <td>14.478041</td>
      <td>11.141163</td>
      <td>11.102427</td>
      <td>11.993050</td>
    </tr>
    <tr>
      <th>Samp_355</th>
      <td>15.997576</td>
      <td>15.448951</td>
      <td>15.355566</td>
      <td>16.501752</td>
      <td>11.701778</td>
      <td>11.323521</td>
      <td>11.610163</td>
      <td>12.733208</td>
      <td>13.650662</td>
      <td>10.770903</td>
      <td>...</td>
      <td>8.711232</td>
      <td>6.977555</td>
      <td>14.805837</td>
      <td>12.283549</td>
      <td>10.847157</td>
      <td>9.833187</td>
      <td>14.037476</td>
      <td>11.082880</td>
      <td>11.708466</td>
      <td>10.654141</td>
    </tr>
    <tr>
      <th>Samp_356</th>
      <td>15.206862</td>
      <td>14.395378</td>
      <td>16.218001</td>
      <td>16.044955</td>
      <td>13.650741</td>
      <td>11.057462</td>
      <td>13.324846</td>
      <td>13.070428</td>
      <td>16.109746</td>
      <td>11.253359</td>
      <td>...</td>
      <td>7.497926</td>
      <td>6.840175</td>
      <td>15.048896</td>
      <td>12.615673</td>
      <td>10.583973</td>
      <td>12.708186</td>
      <td>16.325254</td>
      <td>10.789635</td>
      <td>10.830833</td>
      <td>10.983455</td>
    </tr>
    <tr>
      <th>Samp_357</th>
      <td>14.474129</td>
      <td>15.482863</td>
      <td>15.512549</td>
      <td>15.136613</td>
      <td>14.531277</td>
      <td>10.713539</td>
      <td>11.632242</td>
      <td>13.792981</td>
      <td>13.486404</td>
      <td>10.870028</td>
      <td>...</td>
      <td>7.902283</td>
      <td>7.339226</td>
      <td>14.295151</td>
      <td>11.853855</td>
      <td>10.439473</td>
      <td>11.897782</td>
      <td>16.684790</td>
      <td>10.847185</td>
      <td>11.491449</td>
      <td>11.684467</td>
    </tr>
    <tr>
      <th>Samp_358</th>
      <td>15.094188</td>
      <td>16.047304</td>
      <td>15.298871</td>
      <td>17.022665</td>
      <td>15.046043</td>
      <td>10.931386</td>
      <td>12.289466</td>
      <td>14.327210</td>
      <td>13.533348</td>
      <td>9.792684</td>
      <td>...</td>
      <td>8.017328</td>
      <td>7.814296</td>
      <td>14.267007</td>
      <td>12.117150</td>
      <td>11.832061</td>
      <td>11.021241</td>
      <td>14.533734</td>
      <td>12.015799</td>
      <td>11.551237</td>
      <td>11.221372</td>
    </tr>
  </tbody>
</table>
<p>358 rows × 100 columns</p>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Samp_1</th>
      <td>235.067423</td>
    </tr>
    <tr>
      <th>Samp_2</th>
      <td>253.544991</td>
    </tr>
    <tr>
      <th>Samp_3</th>
      <td>234.204994</td>
    </tr>
    <tr>
      <th>Samp_4</th>
      <td>281.035429</td>
    </tr>
    <tr>
      <th>Samp_5</th>
      <td>245.447781</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>Samp_354</th>
      <td>236.120451</td>
    </tr>
    <tr>
      <th>Samp_355</th>
      <td>222.572359</td>
    </tr>
    <tr>
      <th>Samp_356</th>
      <td>268.472285</td>
    </tr>
    <tr>
      <th>Samp_357</th>
      <td>235.808167</td>
    </tr>
    <tr>
      <th>Samp_358</th>
      <td>213.886123</td>
    </tr>
  </tbody>
</table>
<p>358 rows × 1 columns</p>
</div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Age</th>
      <th>Gender</th>
      <th>BMI</th>
      <th>Chronic_Bronchitis</th>
      <th>Emphysema</th>
      <th>Asthma</th>
    </tr>
    <tr>
      <th>PatientID</th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Samp_1</th>
      <td>78</td>
      <td>0</td>
      <td>31.2</td>
      <td>1</td>
      <td>1</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Samp_2</th>
      <td>68</td>
      <td>1</td>
      <td>19.2</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Samp_3</th>
      <td>54</td>
      <td>1</td>
      <td>19.3</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_4</th>
      <td>47</td>
      <td>1</td>
      <td>36.2</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_5</th>
      <td>60</td>
      <td>1</td>
      <td>26.2</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Samp_354</th>
      <td>71</td>
      <td>0</td>
      <td>23.0</td>
      <td>1</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_355</th>
      <td>62</td>
      <td>1</td>
      <td>25.5</td>
      <td>0</td>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_356</th>
      <td>61</td>
      <td>0</td>
      <td>21.1</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
    <tr>
      <th>Samp_357</th>
      <td>64</td>
      <td>0</td>
      <td>37.6</td>
      <td>0</td>
      <td>0</td>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_358</th>
      <td>61</td>
      <td>1</td>
      <td>31.3</td>
      <td>1</td>
      <td>0</td>
      <td>0</td>
    </tr>
  </tbody>
</table>
<p>358 rows × 6 columns</p>
</div></div>
</div>
</section>
<section id="Generating-a-Multi-Omics-Network-using-SmCCNet">
<h2>Generating a Multi-Omics Network using SmCCNet<a class="headerlink" href="#Generating-a-Multi-Omics-Network-using-SmCCNet" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>SmCCNet stands for: Sparse Multiple Canonical Correlation Network Analysis Tool.</p></li>
<li><p>Is one of our external tools offered through <code class="docutils literal notranslate"><span class="pre">external_tools</span></code>component.</p></li>
<li><p>For more information on SmCCNet please visit docs <a class="reference external" href="https://cran.r-project.org/package=SmCCNet">SmCCNet</a></p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.external_tools</span><span class="w"> </span><span class="kn">import</span> <span class="n">SmCCNet</span>
<span class="n">merged_omics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">omics1</span><span class="p">,</span> <span class="n">omics2</span><span class="p">,</span> <span class="n">clinical</span><span class="p">,</span> <span class="n">phenotype</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">smccnet</span> <span class="o">=</span> <span class="n">SmCCNet</span><span class="p">(</span>
    <span class="n">phenotype_df</span><span class="o">=</span><span class="n">phenotype</span><span class="p">,</span>
    <span class="n">omics_dfs</span><span class="o">=</span><span class="p">[</span><span class="n">omics1</span><span class="p">,</span> <span class="n">omics2</span><span class="p">],</span>
    <span class="n">data_types</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;genes&quot;</span><span class="p">,</span><span class="s2">&quot;proteins&quot;</span><span class="p">],</span>
<span class="p">)</span>
<span class="n">global_network</span><span class="p">,</span> <span class="n">smccnet_clusters</span> <span class="o">=</span> <span class="n">smccnet</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
</section>
<section id="SmCCNet-Output">
<h2>SmCCNet Output<a class="headerlink" href="#SmCCNet-Output" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>SmCCNet returned a 600x600 network and 3 Sub Netwoks or Clusters.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">global_network</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">smccnet_clusters</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(600, 600)
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
3
</pre></div></div>
</div>
</section>
<section id="Generating-Low-Dimensional-Embeddings-using-Graph-Neural-Networks-to-capture-meaningful-biological-interactions.">
<h2>Generating Low-Dimensional Embeddings using Graph Neural Networks to capture meaningful biological interactions.<a class="headerlink" href="#Generating-Low-Dimensional-Embeddings-using-Graph-Neural-Networks-to-capture-meaningful-biological-interactions." title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>This will return node embeddings that reflect both local connectivity and supervised signals</p></li>
<li><p>Each node (omics feature) is associated with a numeric label (e.g., Pearson correlation with phenotype) that guides learning.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span><span class="w"> </span><span class="nn">pandas</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">pd</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.network_embedding</span><span class="w"> </span><span class="kn">import</span> <span class="n">GNNEmbedding</span>

<span class="n">merged_omics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">omics1</span><span class="p">,</span> <span class="n">omics2</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="n">embeddings</span> <span class="o">=</span> <span class="n">GNNEmbedding</span><span class="p">(</span>
    <span class="n">adjacency_matrix</span><span class="o">=</span><span class="n">global_network</span><span class="p">,</span>
    <span class="n">omics_data</span><span class="o">=</span><span class="n">merged_omics</span><span class="p">,</span>
    <span class="n">phenotype_data</span><span class="o">=</span><span class="n">phenotype</span><span class="p">,</span>
    <span class="n">clinical_data</span><span class="o">=</span><span class="n">clinical</span><span class="p">,</span>
    <span class="n">tune</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">embeddings</span><span class="o">.</span><span class="n">fit</span><span class="p">()</span>
<span class="n">embeddings_output</span> <span class="o">=</span> <span class="n">embeddings</span><span class="o">.</span><span class="n">embed</span><span class="p">(</span><span class="n">as_df</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Output">
<h2>Output<a class="headerlink" href="#Output" title="Link to this heading"></a></h2>
<p>For each Omic or Node in the network, our <code class="docutils literal notranslate"><span class="pre">GNNEmbedding</span></code> function generated a 64 dimmensional representation of that Omics.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">embeddings_output</span><span class="o">.</span><span class="n">shape</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
(600, 64)
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_embeddings</span>

<span class="c1"># Using our embeddings instance we get the necessary labels for the graph.</span>
<span class="n">global_node_labels</span> <span class="o">=</span> <span class="n">embeddings</span><span class="o">.</span><span class="n">_prepare_node_labels</span><span class="p">()</span>
<span class="n">embeddings_array</span> <span class="o">=</span> <span class="n">embeddings_output</span><span class="o">.</span><span class="n">values</span>
</pre></div>
</div>
</div>
</section>
<section id="Embeddings-visualization">
<h2>Embeddings visualization<a class="headerlink" href="#Embeddings-visualization" title="Link to this heading"></a></h2>
<p>By visulizing the Emebedding Space in a 2 dimensional space. We can notice some cluster of Nodes/Omics. Highlighting close relationships between them.</p>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">fig1</span> <span class="o">=</span> <span class="n">plot_embeddings</span><span class="p">(</span><span class="n">embeddings_array</span><span class="p">,</span> <span class="n">global_node_labels</span><span class="o">.</span><span class="n">to_frame</span><span class="p">(</span><span class="n">name</span><span class="o">=</span><span class="s2">&quot;phenotype&quot;</span><span class="p">),</span> <span class="n">method</span><span class="o">=</span><span class="s2">&quot;tsne&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/TOPMED_18_0.png" src="_images/TOPMED_18_0.png" />
</div>
</div>
</section>
<section id="Using-the-Embeddings">
<h2>Using the Embeddings<a class="headerlink" href="#Using-the-Embeddings" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>Let use this omics to enrich the representation of the original dataset.</p></li>
<li><p>The <code class="docutils literal notranslate"><span class="pre">SubjectRepresentation</span></code> function takes our previously generated embeddings and our original Omics Dataset and associated Phenotype</p></li>
<li><p>This function will use the embeddings to enrich the orignal dataset. For more details and how this is performed please view our <code class="docutils literal notranslate"><span class="pre">GNN</span> <span class="pre">Embeddings</span> <span class="pre">for</span> <span class="pre">Multi-Omics</span></code> tab.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.downstream_task</span><span class="w">  </span><span class="kn">import</span> <span class="n">SubjectRepresentation</span>

<span class="n">enhanced_omics</span> <span class="o">=</span> <span class="n">SubjectRepresentation</span><span class="p">(</span>
    <span class="n">omics_data</span><span class="o">=</span><span class="n">merged_omics</span><span class="p">,</span>
    <span class="n">embeddings</span><span class="o">=</span><span class="n">embeddings_output</span><span class="p">,</span>
    <span class="n">phenotype_data</span><span class="o">=</span><span class="n">phenotype</span><span class="p">,</span>
    <span class="n">tune</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">enhanced_omics_df</span> <span class="o">=</span> <span class="n">enhanced_omics</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">enhanced_omics_df</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Gene_45</th>
      <th>Gene_197</th>
      <th>Gene_493</th>
      <th>Gene_432</th>
      <th>Gene_357</th>
      <th>Gene_382</th>
      <th>Gene_152</th>
      <th>Gene_427</th>
      <th>Gene_41</th>
      <th>Gene_130</th>
      <th>...</th>
      <th>Gene_202</th>
      <th>Gene_249</th>
      <th>Gene_280</th>
      <th>Gene_62</th>
      <th>Mir_21</th>
      <th>Gene_495</th>
      <th>Mir_59</th>
      <th>Gene_340</th>
      <th>Gene_291</th>
      <th>Gene_81</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Samp_1</th>
      <td>-2.917084</td>
      <td>-13.297334</td>
      <td>-30.571794</td>
      <td>-29.143930</td>
      <td>42.021213</td>
      <td>-12.279085</td>
      <td>-6.925227</td>
      <td>-2.520610</td>
      <td>10.380707</td>
      <td>-6.079539</td>
      <td>...</td>
      <td>10.465338</td>
      <td>1.922512</td>
      <td>5.955446</td>
      <td>-2.626381</td>
      <td>13.947213</td>
      <td>-2.806483</td>
      <td>-0.553959</td>
      <td>1.726373</td>
      <td>-2.651651</td>
      <td>-2.406555</td>
    </tr>
    <tr>
      <th>Samp_2</th>
      <td>-3.102466</td>
      <td>-12.954265</td>
      <td>-28.690063</td>
      <td>-27.614380</td>
      <td>41.053950</td>
      <td>-12.463960</td>
      <td>-6.747382</td>
      <td>-2.712280</td>
      <td>10.841243</td>
      <td>-6.340821</td>
      <td>...</td>
      <td>10.551755</td>
      <td>1.800501</td>
      <td>5.912429</td>
      <td>-2.548239</td>
      <td>12.697251</td>
      <td>-2.648551</td>
      <td>-0.624136</td>
      <td>1.708398</td>
      <td>-2.596500</td>
      <td>-2.342931</td>
    </tr>
    <tr>
      <th>Samp_3</th>
      <td>-2.837912</td>
      <td>-13.429165</td>
      <td>-29.601810</td>
      <td>-28.448539</td>
      <td>39.378358</td>
      <td>-12.837721</td>
      <td>-6.436741</td>
      <td>-2.377240</td>
      <td>10.565243</td>
      <td>-6.044507</td>
      <td>...</td>
      <td>10.307515</td>
      <td>1.991552</td>
      <td>6.038545</td>
      <td>-2.479833</td>
      <td>13.769784</td>
      <td>-2.652630</td>
      <td>-0.592947</td>
      <td>1.583839</td>
      <td>-2.653498</td>
      <td>-2.446277</td>
    </tr>
    <tr>
      <th>Samp_4</th>
      <td>-2.859205</td>
      <td>-12.821326</td>
      <td>-26.844345</td>
      <td>-29.029980</td>
      <td>36.148898</td>
      <td>-12.449477</td>
      <td>-7.229847</td>
      <td>-2.609629</td>
      <td>11.160280</td>
      <td>-6.466545</td>
      <td>...</td>
      <td>9.915145</td>
      <td>2.137159</td>
      <td>5.804663</td>
      <td>-2.629407</td>
      <td>13.458064</td>
      <td>-2.546526</td>
      <td>-0.499689</td>
      <td>1.648553</td>
      <td>-2.520416</td>
      <td>-2.530145</td>
    </tr>
    <tr>
      <th>Samp_5</th>
      <td>-2.982405</td>
      <td>-12.612553</td>
      <td>-29.984057</td>
      <td>-29.110915</td>
      <td>46.903902</td>
      <td>-12.758489</td>
      <td>-6.868394</td>
      <td>-2.698423</td>
      <td>10.211682</td>
      <td>-6.265595</td>
      <td>...</td>
      <td>10.592827</td>
      <td>2.231166</td>
      <td>5.314841</td>
      <td>-2.462155</td>
      <td>12.673634</td>
      <td>-2.597200</td>
      <td>-0.638230</td>
      <td>1.709379</td>
      <td>-2.481453</td>
      <td>-2.397491</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>Samp_354</th>
      <td>-2.935778</td>
      <td>-13.323266</td>
      <td>-29.435019</td>
      <td>-27.890044</td>
      <td>43.240876</td>
      <td>-12.561521</td>
      <td>-6.723627</td>
      <td>-2.290536</td>
      <td>10.847353</td>
      <td>-6.341079</td>
      <td>...</td>
      <td>10.186933</td>
      <td>2.077744</td>
      <td>5.888022</td>
      <td>-2.566317</td>
      <td>13.658852</td>
      <td>-2.439809</td>
      <td>-0.526501</td>
      <td>1.692860</td>
      <td>-2.711183</td>
      <td>-2.466377</td>
    </tr>
    <tr>
      <th>Samp_355</th>
      <td>-3.065084</td>
      <td>-13.295135</td>
      <td>-30.718263</td>
      <td>-29.112126</td>
      <td>44.613763</td>
      <td>-12.331960</td>
      <td>-6.955723</td>
      <td>-2.971791</td>
      <td>10.294231</td>
      <td>-6.077254</td>
      <td>...</td>
      <td>9.624207</td>
      <td>2.077401</td>
      <td>5.623264</td>
      <td>-2.663072</td>
      <td>14.311062</td>
      <td>-2.595351</td>
      <td>-0.584500</td>
      <td>1.673485</td>
      <td>-2.578159</td>
      <td>-2.418925</td>
    </tr>
    <tr>
      <th>Samp_356</th>
      <td>-3.069497</td>
      <td>-14.104972</td>
      <td>-29.862862</td>
      <td>-30.838439</td>
      <td>39.916304</td>
      <td>-12.129115</td>
      <td>-6.739620</td>
      <td>-2.504380</td>
      <td>10.865619</td>
      <td>-6.296479</td>
      <td>...</td>
      <td>10.497804</td>
      <td>1.984929</td>
      <td>5.467676</td>
      <td>-2.442813</td>
      <td>14.854524</td>
      <td>-2.540575</td>
      <td>-0.527353</td>
      <td>1.648124</td>
      <td>-2.674988</td>
      <td>-2.482001</td>
    </tr>
    <tr>
      <th>Samp_357</th>
      <td>-3.112670</td>
      <td>-14.041763</td>
      <td>-28.716949</td>
      <td>-28.512552</td>
      <td>48.270441</td>
      <td>-12.605819</td>
      <td>-6.424579</td>
      <td>-2.768089</td>
      <td>11.026826</td>
      <td>-6.542514</td>
      <td>...</td>
      <td>10.566136</td>
      <td>2.141326</td>
      <td>6.178911</td>
      <td>-2.667312</td>
      <td>13.193234</td>
      <td>-2.821949</td>
      <td>-0.585463</td>
      <td>1.659515</td>
      <td>-2.625582</td>
      <td>-2.425626</td>
    </tr>
    <tr>
      <th>Samp_358</th>
      <td>-2.983813</td>
      <td>-13.657581</td>
      <td>-28.264297</td>
      <td>-28.283759</td>
      <td>41.077029</td>
      <td>-12.508963</td>
      <td>-6.906523</td>
      <td>-2.443945</td>
      <td>11.551972</td>
      <td>-6.444730</td>
      <td>...</td>
      <td>10.345094</td>
      <td>2.132126</td>
      <td>5.578996</td>
      <td>-2.523854</td>
      <td>14.942860</td>
      <td>-2.537228</td>
      <td>-0.534264</td>
      <td>1.689256</td>
      <td>-2.751487</td>
      <td>-2.255327</td>
    </tr>
  </tbody>
</table>
<p>358 rows × 600 columns</p>
</div></div>
</div>
</section>
<section id="Comparing-results">
<h2>Comparing results<a class="headerlink" href="#Comparing-results" title="Link to this heading"></a></h2>
<p>Lets compare the enriched dataset vs the raw dataset by using it with a popular machine learning technique, <code class="docutils literal notranslate"><span class="pre">Random</span> <span class="pre">Forest</span></code>. In this example we are simply using the high-dimensional omics data to make prediction on the phenotype.</p>
<ul class="simple">
<li><p>Our entire codebase is publicly avaible at the <a class="reference external" href="https://github.com/UCD-BDLab">Big Data Management and Mining Laboratory</a> github repository.</p></li>
<li><p>For specific details on this code, please visit: <a class="reference external" href="https://github.com/UCD-BDLab/BioNeuralNet">BioNeuralNet</a></p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">evaluate_rf</span><span class="p">,</span> <span class="n">plot_performance</span>

<span class="n">y_phenotype</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">values</span>
<span class="n">X_enriched_omics</span> <span class="o">=</span> <span class="n">enhanced_omics_df</span><span class="o">.</span><span class="n">values</span>
<span class="n">X_raw_omics</span> <span class="o">=</span> <span class="n">merged_omics</span><span class="o">.</span><span class="n">values</span>

<span class="n">accuracy_with_embeddings</span> <span class="o">=</span> <span class="n">evaluate_rf</span><span class="p">(</span><span class="n">X_enriched_omics</span><span class="p">,</span> <span class="n">y_phenotype</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;regression&quot;</span><span class="p">)</span>
<span class="n">accuracy_alone</span> <span class="o">=</span> <span class="n">evaluate_rf</span><span class="p">(</span><span class="n">X_enriched_omics</span><span class="p">,</span> <span class="n">y_phenotype</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;regression&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">plot_performance</span><span class="p">(</span><span class="n">accuracy_with_embeddings</span><span class="p">,</span> <span class="n">accuracy_alone</span><span class="p">,</span> <span class="s2">&quot;Raw Omics vs Enriched Omics&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/TOPMED_24_0.png" src="_images/TOPMED_24_0.png" />
</div>
</div>
</section>
<section id="Network-Visulization">
<h2>Network Visulization<a class="headerlink" href="#Network-Visulization" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>As part of this demonstration, I have developed a number of graphic and plotting tools.</p></li>
<li><p>Please note that these tools are not part of the BioNeuralNet core, and it is not my intent to present them this way.</p></li>
<li><p>This examples highlight how external libraies and your own code can be easily integrated into our workflow. Allowing users to further explore the omics-data.</p></li>
<li><p>All this code and visulization aid components will be availble after the presentation.</p></li>
</ul>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Lets take a quick look at some of those clusters from SmCCNet</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_network</span>

<span class="n">cluster1</span><span class="p">,</span> <span class="n">cluster2</span><span class="p">,</span> <span class="n">cluster3</span> <span class="o">=</span> <span class="n">smccnet_clusters</span>

<span class="n">cluster1_mapping</span> <span class="o">=</span> <span class="n">plot_network</span><span class="p">(</span><span class="n">cluster2</span><span class="p">,</span> <span class="n">weight_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">show_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_edge_weights</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">cluster1_mapping</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>

<span class="n">cluster2_mapping</span> <span class="o">=</span> <span class="n">plot_network</span><span class="p">(</span><span class="n">cluster1</span><span class="p">,</span> <span class="n">weight_threshold</span><span class="o">=</span><span class="mf">0.009</span><span class="p">,</span> <span class="n">show_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_edge_weights</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">cluster2_mapping</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/TOPMED_26_0.png" src="_images/TOPMED_26_0.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Omic</th>
      <th>Degree</th>
    </tr>
    <tr>
      <th>Index</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>15</th>
      <td>Mir_2</td>
      <td>14</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Gene_1</td>
      <td>4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Gene_2</td>
      <td>4</td>
    </tr>
    <tr>
      <th>3</th>
      <td>Gene_6</td>
      <td>4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Gene_7</td>
      <td>4</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/TOPMED_26_2.png" src="_images/TOPMED_26_2.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Omic</th>
      <th>Degree</th>
    </tr>
    <tr>
      <th>Index</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>7</th>
      <td>Mir_53</td>
      <td>6</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Gene_5</td>
      <td>4</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Gene_54</td>
      <td>4</td>
    </tr>
    <tr>
      <th>4</th>
      <td>Gene_88</td>
      <td>3</td>
    </tr>
    <tr>
      <th>5</th>
      <td>Gene_174</td>
      <td>3</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
</section>
<section id="Correlated-Clustering">
<h2>Correlated Clustering<a class="headerlink" href="#Correlated-Clustering" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>BioNeuralNet includes internal modules for performing correlated clustering on complex networks.</p></li>
<li><p>These methods modify and extend the traditional community detection by integrating phenotype correlation, allowing users to extract biologically relevant, phenotype-associated modules from any network.</p></li>
<li><p>For more details on how this performed, please visit our <code class="docutils literal notranslate"><span class="pre">Correlated</span> <span class="pre">Clustering</span> <span class="pre">Methods</span></code> tab</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.clustering</span><span class="w"> </span><span class="kn">import</span> <span class="n">CorrelatedLouvain</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">networkx</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">nx</span>

<span class="n">merged_omics</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">omics1</span><span class="p">,</span> <span class="n">omics2</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
<span class="n">G_network</span> <span class="o">=</span> <span class="n">nx</span><span class="o">.</span><span class="n">from_pandas_adjacency</span><span class="p">(</span><span class="n">global_network</span><span class="p">)</span>

<span class="n">louvain_instance</span> <span class="o">=</span> <span class="n">CorrelatedLouvain</span><span class="p">(</span>
    <span class="n">G</span><span class="o">=</span><span class="n">G_network</span><span class="p">,</span>
    <span class="n">B</span><span class="o">=</span><span class="n">merged_omics</span><span class="p">,</span>
    <span class="n">Y</span><span class="o">=</span><span class="n">phenotype</span><span class="p">,</span>
    <span class="n">tune</span><span class="o">=</span><span class="kc">True</span>
<span class="p">)</span>

<span class="n">louvain_clusters</span> <span class="o">=</span> <span class="n">louvain_instance</span><span class="o">.</span><span class="n">run</span><span class="p">(</span><span class="n">as_dfs</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="nb">print</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">louvain_clusters</span><span class="p">))</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
3
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Lets comapre these clusters against SmCCNet clusters</span>
<span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">compare_clusters</span>

<span class="n">compare_clusters</span><span class="p">(</span><span class="n">louvain_clusters</span><span class="p">,</span> <span class="n">smccnet_clusters</span><span class="p">,</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">merged_omics</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/TOPMED_30_0.png" src="_images/TOPMED_30_0.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Cluster</th>
      <th>Louvain Size</th>
      <th>Louvain Correlation</th>
      <th>SMCCNET Size</th>
      <th>SMCCNET Correlation</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>Cluster_1</td>
      <td>77</td>
      <td>0.724029</td>
      <td>67</td>
      <td>0.134853</td>
    </tr>
    <tr>
      <th>1</th>
      <td>Cluster_2</td>
      <td>76</td>
      <td>-0.124891</td>
      <td>63</td>
      <td>0.731769</td>
    </tr>
    <tr>
      <th>2</th>
      <td>Cluster_3</td>
      <td>6</td>
      <td>0.069064</td>
      <td>15</td>
      <td>-0.085057</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_network</span><span class="p">,</span> <span class="n">louvain_to_adjacency</span>

<span class="n">cluster1</span> <span class="o">=</span> <span class="n">louvain_clusters</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
<span class="n">cluster2</span> <span class="o">=</span> <span class="n">louvain_clusters</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span>

<span class="c1"># Convert Louvain clusters into adjacency matrices</span>
<span class="n">louvain_adj1</span> <span class="o">=</span> <span class="n">louvain_to_adjacency</span><span class="p">(</span><span class="n">cluster1</span><span class="p">)</span>
<span class="n">louvain_adj2</span> <span class="o">=</span> <span class="n">louvain_to_adjacency</span><span class="p">(</span><span class="n">cluster2</span><span class="p">)</span>

<span class="c1"># Plot using the converted adjacency matrices</span>
<span class="n">cluster1_mapping</span> <span class="o">=</span> <span class="n">plot_network</span><span class="p">(</span><span class="n">louvain_adj1</span><span class="p">,</span> <span class="n">weight_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">show_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_edge_weights</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">cluster1_mapping</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>

<span class="n">cluster2_mapping</span> <span class="o">=</span> <span class="n">plot_network</span><span class="p">(</span><span class="n">louvain_adj2</span><span class="p">,</span> <span class="n">weight_threshold</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span> <span class="n">show_labels</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">show_edge_weights</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">cluster2_mapping</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/TOPMED_31_0.png" src="_images/TOPMED_31_0.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Omic</th>
      <th>Degree</th>
    </tr>
    <tr>
      <th>Index</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>34</th>
      <td>Gene_222</td>
      <td>7</td>
    </tr>
    <tr>
      <th>30</th>
      <td>Mir_2</td>
      <td>7</td>
    </tr>
    <tr>
      <th>66</th>
      <td>Gene_1</td>
      <td>7</td>
    </tr>
    <tr>
      <th>45</th>
      <td>Gene_6</td>
      <td>6</td>
    </tr>
    <tr>
      <th>53</th>
      <td>Gene_7</td>
      <td>6</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/TOPMED_31_2.png" src="_images/TOPMED_31_2.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Omic</th>
      <th>Degree</th>
    </tr>
    <tr>
      <th>Index</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>30</th>
      <td>Gene_187</td>
      <td>9</td>
    </tr>
    <tr>
      <th>44</th>
      <td>Gene_464</td>
      <td>6</td>
    </tr>
    <tr>
      <th>21</th>
      <td>Gene_83</td>
      <td>6</td>
    </tr>
    <tr>
      <th>16</th>
      <td>Gene_354</td>
      <td>6</td>
    </tr>
    <tr>
      <th>9</th>
      <td>Gene_300</td>
      <td>5</td>
    </tr>
  </tbody>
</table>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.datasets</span><span class="w"> </span><span class="kn">import</span> <span class="n">DatasetLoader</span>
<span class="kn">import</span><span class="w"> </span><span class="nn">numpy</span><span class="w"> </span><span class="k">as</span><span class="w"> </span><span class="nn">np</span>

<span class="n">loader</span> <span class="o">=</span> <span class="n">DatasetLoader</span><span class="p">(</span><span class="s2">&quot;example1&quot;</span><span class="p">)</span>
<span class="n">omics1</span><span class="p">,</span> <span class="n">omics2</span><span class="p">,</span> <span class="n">phenotype</span><span class="p">,</span> <span class="n">clinical</span> <span class="o">=</span> <span class="n">loader</span><span class="o">.</span><span class="n">load_data</span><span class="p">()</span>
<span class="n">display</span><span class="p">(</span><span class="n">phenotype</span><span class="p">)</span>

<span class="n">min_val</span> <span class="o">=</span> <span class="n">phenotype</span><span class="p">[</span><span class="s2">&quot;phenotype&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">min</span><span class="p">()</span>
<span class="n">max_val</span> <span class="o">=</span> <span class="n">phenotype</span><span class="p">[</span><span class="s2">&quot;phenotype&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">max</span><span class="p">()</span>

<span class="c1"># linspace creates an array of evenly spaced values</span>
<span class="n">bins</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">linspace</span><span class="p">(</span><span class="n">min_val</span><span class="p">,</span> <span class="n">max_val</span><span class="p">,</span> <span class="mi">5</span><span class="p">)</span>

<span class="n">phenotype</span><span class="p">[</span><span class="s2">&quot;phenotype&quot;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">cut</span><span class="p">(</span><span class="n">phenotype</span><span class="p">[</span><span class="s2">&quot;phenotype&quot;</span><span class="p">],</span> <span class="n">bins</span><span class="o">=</span><span class="n">bins</span><span class="p">,</span> <span class="n">labels</span><span class="o">=</span><span class="p">[</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">3</span><span class="p">],</span> <span class="n">include_lowest</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
<span class="n">count_values</span> <span class="o">=</span> <span class="n">phenotype</span><span class="p">[</span><span class="s2">&quot;phenotype&quot;</span><span class="p">]</span><span class="o">.</span><span class="n">value_counts</span><span class="p">(</span><span class="n">sort</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>

<span class="n">display</span><span class="p">(</span><span class="n">phenotype</span><span class="p">)</span>
<span class="n">display</span><span class="p">(</span><span class="n">count_values</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Samp_1</th>
      <td>235.067423</td>
    </tr>
    <tr>
      <th>Samp_2</th>
      <td>253.544991</td>
    </tr>
    <tr>
      <th>Samp_3</th>
      <td>234.204994</td>
    </tr>
    <tr>
      <th>Samp_4</th>
      <td>281.035429</td>
    </tr>
    <tr>
      <th>Samp_5</th>
      <td>245.447781</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>Samp_354</th>
      <td>236.120451</td>
    </tr>
    <tr>
      <th>Samp_355</th>
      <td>222.572359</td>
    </tr>
    <tr>
      <th>Samp_356</th>
      <td>268.472285</td>
    </tr>
    <tr>
      <th>Samp_357</th>
      <td>235.808167</td>
    </tr>
    <tr>
      <th>Samp_358</th>
      <td>213.886123</td>
    </tr>
  </tbody>
</table>
<p>358 rows × 1 columns</p>
</div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>phenotype</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Samp_1</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_2</th>
      <td>2</td>
    </tr>
    <tr>
      <th>Samp_3</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_4</th>
      <td>3</td>
    </tr>
    <tr>
      <th>Samp_5</th>
      <td>2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
    </tr>
    <tr>
      <th>Samp_354</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_355</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_356</th>
      <td>2</td>
    </tr>
    <tr>
      <th>Samp_357</th>
      <td>1</td>
    </tr>
    <tr>
      <th>Samp_358</th>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>358 rows × 1 columns</p>
</div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
phenotype
0     38
1    158
2    141
3     21
Name: count, dtype: int64
</pre></div></div>
</div>
</section>
<section id="DPMON-(Disease-Prediction-using-Multi-Omics-Networks)-reuses-the-same-GNN-architectures-but-with-a-different-objective:">
<h2>DPMON (Disease Prediction using Multi-Omics Networks) reuses the same GNN architectures but with a different objective:<a class="headerlink" href="#DPMON-(Disease-Prediction-using-Multi-Omics-Networks)-reuses-the-same-GNN-architectures-but-with-a-different-objective:" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p>DPMON aggregates node embeddings with patient-level omics data.</p></li>
<li><p>A downstream classification head is applied for sample-level disease prediction.</p></li>
<li><p>This end-to-end approach leverages both local (node-level) and global (patient-level) network information.</p></li>
<li><p>This single cell bellow captures the entire workflow demonstrated earlier (Generating GNN Embeddins + Integrating these Embeddings back into Omics Dataset) in an end-to-end iterative pipeline.</p></li>
</ul>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.downstream_task</span><span class="w"> </span><span class="kn">import</span> <span class="n">DPMON</span>
<span class="n">dpmon</span> <span class="o">=</span> <span class="n">DPMON</span><span class="p">(</span>
    <span class="n">adjacency_matrix</span><span class="o">=</span><span class="n">global_network</span><span class="p">,</span>
    <span class="n">omics_list</span><span class="o">=</span><span class="p">[</span><span class="n">omics1</span><span class="p">,</span> <span class="n">omics2</span><span class="p">],</span>
    <span class="n">phenotype_data</span><span class="o">=</span><span class="n">phenotype</span><span class="p">,</span>
    <span class="n">clinical_data</span><span class="o">=</span><span class="n">clinical</span><span class="p">,</span>
    <span class="n">tune</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
<span class="p">)</span>
<span class="n">dpmon_predictions</span> <span class="o">=</span> <span class="n">dpmon</span><span class="o">.</span><span class="n">run</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">display</span><span class="p">(</span><span class="n">dpmon_predictions</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Actual</th>
      <th>Predicted</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>1</th>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>3</th>
      <td>3</td>
      <td>3</td>
    </tr>
    <tr>
      <th>4</th>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>353</th>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>354</th>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>355</th>
      <td>2</td>
      <td>2</td>
    </tr>
    <tr>
      <th>356</th>
      <td>1</td>
      <td>1</td>
    </tr>
    <tr>
      <th>357</th>
      <td>1</td>
      <td>1</td>
    </tr>
  </tbody>
</table>
<p>358 rows × 2 columns</p>
</div></div>
</div>
</section>
<section id="DPMON-allows-BioNeuralNet-users-to-significatly-improve-phenotype-predictions-with-a-few-lines-of-code.">
<h2>DPMON allows BioNeuralNet users to significatly improve phenotype predictions with a few lines of code.<a class="headerlink" href="#DPMON-allows-BioNeuralNet-users-to-significatly-improve-phenotype-predictions-with-a-few-lines-of-code." title="Link to this heading"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[ ]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.metrics</span><span class="w"> </span><span class="kn">import</span> <span class="n">evaluate_rf</span>

<span class="n">X_raw</span> <span class="o">=</span> <span class="n">merged_omics</span><span class="o">.</span><span class="n">values</span>
<span class="n">y_global</span> <span class="o">=</span> <span class="n">phenotype</span><span class="o">.</span><span class="n">values</span>
<span class="n">raw_rf_acc</span> <span class="o">=</span> <span class="n">evaluate_rf</span><span class="p">(</span><span class="n">X_raw</span><span class="p">,</span> <span class="n">y_global</span><span class="p">,</span> <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;classification&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="kn">from</span><span class="w"> </span><span class="nn">bioneuralnet.utils</span><span class="w"> </span><span class="kn">import</span> <span class="n">plot_performance</span>
<span class="n">plot_performance</span><span class="p">(</span><span class="n">dpmon_predictions</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="n">raw_rf_acc</span><span class="p">,</span> <span class="s2">&quot;Raw Omics, vs DPMON Omics&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="_images/TOPMED_38_0.png" src="_images/TOPMED_38_0.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="TCGA-BRCA_Dataset.html" class="btn btn-neutral float-left" title="TCGA-BRCA Dataset" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="tutorials/index.html" class="btn btn-neutral float-right" title="Tutorials" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright .</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>